File: examples/Python/asyncio_ioloop/tasksink.py
#!/usr/bin/env python

"""
synopsis:
    Task sink
    Binds PULL socket to tcp://localhost:5558
    Collects results from workers via that socket
    Author: Lev Givon <lev(at)columbia(dot)edu>
    Modified for async/ioloop: Dave Kuhlman <dkuhlman(at)davekuhlman(dot)org>
usage:
    python tasksink.py
"""

import sys
import time
import zmq
from zmq.asyncio import Context, ZMQEventLoop
import asyncio


@asyncio.coroutine
def run_sink(context):
    # Socket to receive messages on
    receiver = context.socket(zmq.PULL)
    receiver.bind("tcp://*:5558")
    # Wait for start of batch
    yield from receiver.recv()
    # Start our clock now
    tstart = time.time()
    # Process 100 confirmations
    for task_nbr in range(100):
        yield from receiver.recv()
        if task_nbr % 10 == 0:
            sys.stdout.write(':')
        else:
            sys.stdout.write('.')
        sys.stdout.flush()
    # Calculate and report duration of batch
    tend = time.time()
    print("Total elapsed time: %d msec" % ((tend - tstart) * 1000))


@asyncio.coroutine
def run(loop):
    context = Context()
    yield from run_sink(context)


def main():
    args = sys.argv[1:]
    if len(args) != 0:
        sys.exit(__doc__)
    try:
        loop = ZMQEventLoop()
        asyncio.set_event_loop(loop)
        loop.run_until_complete(run(loop))
    except KeyboardInterrupt:
        print('\nFinished (interrupted)')


if __name__ == '__main__':
    main()

File: examples/Java/src/main/java/com/imatix/zguide/ex1_10/TaskSink.java
package com.imatix.zguide.ex1_10;

import org.zeromq.ZMQ.Context;
import org.zeromq.ZMQ.Socket;

import static com.imatix.zguide.commons.ZConstants.URL.TCP_5558;
import static java.lang.System.currentTimeMillis;
import static org.zeromq.ZMQ.PULL;
import static org.zeromq.ZMQ.context;

/**
 * Task sink in Java
 * Binds PULL socket to tcp://localhost:5558
 * Collects results from workers via that socket
 *
 * @since 1.0
 */

public class TaskSink {

    public static void main(String[] args) {
        //  Prepare our context and socket
        Context context = context(1);
        Socket receiver = context.socket(PULL);
        receiver.bind(TCP_5558);
        //  Wait for start of batch
        String message = new String(receiver.recv(0));
        System.out.println("Start of the batch: " + message);
        //  Start our clock now
        long start = currentTimeMillis();
        //  Process 100 confirmations
        int taskNum;
        for (taskNum = 0; taskNum < 100; taskNum++) {
            new String(receiver.recv(0));
            if ((taskNum / 10) * 10 == taskNum)
                System.out.println(":");
            else
                System.out.println(".");
        }
        //  Calculate and report duration of batch
        long end = currentTimeMillis();
        System.out.printf("\nTotal elapsed time: %d msec.", (end - start));
        receiver.close();
        context.term();
    }
}
